util/db/dbutil.go

```diff
@@ -219,6 +219,14 @@ func (db *Accessor) Atomic(fn idemFn, extras ...interface{}) (err error) {
 // For transactions where readOnly is false, sync determines whether or not to wait for the result.
 // Like for Atomic, the return error of fn should be a native sqlite3.Error type or an error wrapping it.
 func (db *Accessor) AtomicContext(ctx context.Context, fn idemFn, extras ...interface{}) (err error) {
+
+	return db.AtomicContextWithRetryClearFn(ctx, fn, nil, extras...)
+}
+
+// AtomicContextWithRetryClearFn is like AtomicContext, but calls retryClearFn if the database
+// txn was rolled back, due to error or in between retries. This helps a caller that
+// might change in-memory state inside fn.
+func (db *Accessor) AtomicContextWithRetryClearFn(ctx context.Context, fn idemFn, retryClearFn func(context.Context), extras ...interface{}) (err error) {
```

Comment:
Kind of wondering if we should just have this function/remove AtomicContext standalone, but that doesn't have to block this PR.